mars/services/task/supervisor/manager.py

```diff
@@ -174,12 +161,61 @@ async def submit_tileable_graph(
             self._task_preprocessor_cls,
         )
 
+        def _on_finalize():
+            # The loop may be closed before the weakref is dead.
+            if loop.is_running():
+                aiotask = loop.create_task(
+                    self._move_task_to_reserved(loop, task_id, processor_ref)
+                )
+                add_aiotask_done_check_callback(
+                    aiotask,
+                    "Move task to reserved failed, task id: %s, processor ref: %s",
+                    task_id,
+                    processor_ref,
+                )
+
+        loop = asyncio.get_running_loop()
+        task_ref = _RefHolder()
+        weakref.finalize(task_ref, _on_finalize)
         for tileable in graph.result_tileables:
-            info = ResultTileableInfo(tileable=tileable, processor_ref=processor_ref)
-            self._tileable_key_to_info[tileable.key].append(info)
+            info = ResultTileableInfo(
+                tileable=tileable, processor_ref=processor_ref, ref_holder=task_ref
+            )
+            logger.debug(
+                "Add tileable info, task id: %s, tileable key: %s",
+                task_id,
+                tileable.key,
+            )
+            self._result_tileable_key_to_info[tileable.key].append(info)
 
         return task_id
 
+    async def _move_task_to_reserved(self, loop, task_id, processor_ref):
+        # TODO(fyrestone): Find a better way to wait and destroy the processor actor.
+        with contextlib.suppress(ActorNotExist, ServerClosed, ConnectionRefusedError):
+            await processor_ref.wait()
+
+        logger.debug("Move task %s to reserved.", task_id)
+        ref_holder = _RefHolder()
+        self._reserved_finish_tasks.append(ref_holder)
+
+        async def _destroy_actor():
+            with contextlib.suppress(
+                ActorNotExist, ServerClosed, ConnectionRefusedError
+            ):
+                await processor_ref.destroy()
+
+        def _remove_task():
+            logger.debug("Remove task %s.", task_id)
+            self._task_id_to_processor_ref.pop(task_id, None)
+            if loop.is_running():
+                aiotask = loop.create_task(_destroy_actor())
+                add_aiotask_done_check_callback(
+                    aiotask, "Destroy processor ref %s failed.", processor_ref
+                )
+
+        weakref.finalize(ref_holder, _remove_task)
```

Comment:
Can we just check whether `self._reserved_finish_tasks.append(ref_holder)` pop item, and execute `_remove_task` body directly?

Reply:
The `_remove_task` and `_destroy_actor` are always needed to be called. But, poping queue manually should handle two cases,
1, The queue size > 0, pop first, call `_remove_task`, then push last.
2, The queue size is 0, call `_remove_task` directly, not to push.

Use weakref is a bit simpler than pop queue manually.